### ISSUE TITLE
Push the mutation query into queue when network error occurs but client is not offline?

### ISSUE THREAD

**Issue by skylarkgit** [aerogear/offix#162]
I expected the mutation to be pushed into the offline queue if the device is not offline and a network error occurs. But it just doesn't happen and I can't also find the method to watchOfflineChanges.
```

const mutationOptions = {
    mutation: ADD_TASKS,
    variables: {
        label: "wtf"
    },
    updateQuery: {
        query: ADD_TASKS,
        variables: {
            label: "wtf"
        }
    },
    returnType: "Task",
    operationType: CacheOperation.ADD,
    idField: "id"
};

const client = new OfflineClient(offixConfig);
client.init().then(_ => {
    client
        .offlineMutation(mutationOptions)
        .then(data => console.log("data", data))
        .catch(error => {
            if (error.networkError /*&& error.networkError.offline*/) {
                const offlineError =  error.networkError;
                offlineError.watchOfflineChange().then(console.log("offline_data", data));
            } else {
                console.log("err", error);
            }
            return error;
        });
});
```

I added the offlineQueueListener here:
```
export const offixConfig = {
	httpUrl: "http://192.168.43.39:4000",
	wsUrl: "ws://192.168.43.39:4000",
	offlineQueueListener: data => {
		console.log("queue", data);
	}
};

```

Or may be, am I doing something wrong?

**Comment by wtrocki** [comment]
Hi @skylarkgit 

Thank you for logging this issue. Sorry for getting back to you so late. 
I saw your message earlier but wanted to check something before I write a response

TL;DR There is no support for this out of the box but this is possible now with offix. 

Generally, only a reliable way to determine if the application is offline is to make call to the server to get some result or relly on the operating system to tell you that we are offline - airplane case. Everything between is a little bit grey area. Network error can occur because of different cases. For example timeout can mean that the client did not get a response back - but server might actually process some of the data - doing stuff like that automatically can lead to some issues like duplicates. We had this automatic behavior before and it was leading to issues that our community reported

So offix by default relies on the system network state and provides a way to detect offline specific error on catch.

However, if your server returns an error or any other error we can still support this case by:

#### Enqueue operation as offline

When making a request to server if we get network error we can simply detect 
what kind of error we are dealing with and if we are happy to enqueue this operation as offline we can execute mutation again with the same parameters, 
but this time passes extra flag.

```
context: {...context, isOffline: true } 
```

What I have noticed that using this however will make optimistic response to dissapear after restart. I think we need to have extra parameter.
For example:

```
context: {...context, saveToOffline: true }
```

That will explicitly save offline change. 
This change is very simple however I'm on holidays at the moment so I will be able 
#### Separate link

If we do not care about specifics of network errors and expect things to happen out of the box we can use retry link handler that can be passed along with httpLink to `config.terminatingLink`


https://github.com/carvajalconsultants/apollo-link-offline

This library does that well, however, it will not allow

## Actions from this issue

Fix bug for enqueuing offline operation (ETA 1.5 week to next release)
Document this case
Provide a clear way for developers to enqueue operation again.

**Comment by skylarkgit** [comment]
Thanks, I guess the first solution will work for me.

**Comment by wtrocki** [comment]
@garsharm `context: {...context, isOffline: true } ` works as we using it in tests. Linked PR cannot be merged as we making now refactor to pass every single change thru the queue - so it will not be required. New release should be happening in coming weeks.

**Comment by garsharm** [comment]
Okay, @wtrocki. Thanks for the quick reply. and a few things I want to clarify regarding the implementation.
 1. as of now  `context: {...context, isOffline: true }` is pushing the mutation in the `offlineQueueListener: {
                onOperationEnqueued: (entry) => {
                    console.log(entry)
                }
            }` but not queuing in offline store array. right? I think because we are pending with Issue [#147](https://github.com/aerogear/offix/issues/147). 
2. Also, the functionality to trigger these queued mutations is not implemented yet. right?

**Comment by wtrocki** [comment]
@garsharm We are actively working on a feature that will enable that in https://github.com/aerogear/offix/pull/170 but it will take some time to get it released.

**Comment by darahayes** [comment]
Hey folks, just wanted to jump in here. I don't think I'll be very helpful in providing you a solution that you can use **right now** but I did want to just provide some info on the work I'm doing that may help you in the future.

> I expected the mutation to be pushed into the offline queue if the device is not offline and a network error occurs.

This is an area we've struggled with. In an ideal world we would do this automatically but the challenge is 'How do we define what is and isn't a network error?' As @wtrocki mentioned before, we've tried this in the past and encountered issues. One of the things we want to avoid is having an item in the Offline Queue that simply cannot be fulfilled. If this happens we need extra logic to catch that case, or the queue seizes up and your application gets into a pretty bad state.

The other thing is that different users will have their own opinions and definitions for the various error states that their application will deal with. One example of how we deal with that is the `NetworkStatus` interface, but it seems still have issues around making it easier to decide how/when something should be enqueued.

There have been some suggestions above for extra flags such as `isOffline` and `saveToOffline` but I feel like these flags don't make a lot of sense, and they would only enable really specific use cases. It would be an ok solution until someone else comes along and wants another flag for their particular use case. Next thing we know, the API is bloated with flags that are impossible to understand.

In my opinion, the bigger problem here is that the Offline Queue is too much of a black box. Ultimately I think the queue needs to be accessible so users can decide themselves how/what/when to enqueue operations, and to have some control over when they will be fulfilled.

Right now I'm working on #170 which will give you direct access to the queue on `client.queue`.

Just some examples of things you'll be able to do on the queue itself:

* `queue.enqueueOperation` - add a mutation to the queue (and persist it)
* `queue.dequeueOperation` - remove a mutation from the queue (and remove from storage)
* `queue.forwardOperations` - fulfil all operations in the queue.

I see `client.offlineMutate` as being a high level API that can cover the most common situations and if you need to do some more advanced things, then the lower level `queue` API will be available directly.

Any feedback on those ideas would be massively helpful!

**Comment by garsharm** [comment]
@wtrocki and @darahayes  thank you for the clarification on the issue. I totally agree that API should be easily understandable and accessible.
> Right now I'm working on #170 which will give you direct access to the queue on client.queue.

I will be going through the PR and will be ready to help if needed anywhere. 
Cheers!

**Comment by darahayes** [comment]
> I will be going through the PR and will be ready to help if needed anywhere.

Beware it is in a pretty rough state right now. I'm working on introducing a bunch of tests.

**Comment by garsharm** [comment]
no problem :+1:

**Comment by darahayes** [comment]
Just wanted to post an update here. We've had two releases since this issue was first opened. It should be now possible to requeue the failed operation if that is something you wish to do. As an example you could probably do something like this.

```js
client
  .offlineMutate(mutationOptions)
  .then(data => console.log("data", data))
  .catch(error => {
    if (error.offline) {
	  // error.offline is true if it's been queued by offix
      // in that case `watchOfflineChange()` will be available
      error.watchOfflineChange().then(console.log("offline_data", data));
    } else if (error.networkError) {
      // check some other condition such as networkError or something else
      // enqueue the operation again
      client.queue.enqueueOperation(mutationOptions).then(console.log).catch(console.log)
    }
    console.log("err", error);
    return error;
});
```

Admittedly, the API here isn't great and could possibly be refined. Since the [0.9.0](https://github.com/aerogear/offix/releases/tag/0.9.0) release there is direct access to `client.queue` which should help you get closer to where you want to be. We'd be really happy if you tried it and reported back any feedback/suggestions for how we can further improve things.

**Comment by wtrocki** [comment]
@darahayes Having this comment ported to documentation will be amazing. 
This is very common use case. Separate chapter that will have exactly the same text will do wonders

**Comment by darahayes** [comment]
@wtrocki this does all come back to a topic we've talked about before. Should we schedule every single operation regardless of being online/offline?

I think we need to think about it more seriously and try to understand the challenges. The issue that comes to mind is how to handle a bad operation that will never be fulfilled. We would need some way to flush that automatically or we would need to provide an API to make it possible for developers to do that.

**Comment by wtrocki** [comment]
We had this concept initially that attached numer of unsuccessful requests to operations and allow developers to configure that. We have ported that to retry-link that is being used for scheduler now. Retry link provides great number of defaults for this purpose and it can be configured to encompass a number of the use cases. 

Rescheduling with mixture of the offline and online changes is risky business from conflict side. 
For simplicity we might need to evaluate using scheduler for every operation that is being send to server (synchronously) but this might have serious impact on performance and UX

**Comment by wtrocki** [comment]
TL;DR - Agreed that this needs more oversight to get into docs.

**Comment by wtrocki** [comment]
Refreshing this topic. Generally we need capability to:

- Add item to the queue 
   - Directly in user app space
   - As the form of retry/error link failover
- Give ability to expire items from the queue.
It can be based on
    - Manual trigger 🙈 
    - no of tries
     - expiry time?

CC @darahayes

**Comment by wtrocki** [comment]
While manual handling is also an option it will end up with lots of boilerplate code in react components that we can avoid.

**Comment by darahayes** [comment]
Ultimately this is a weakness in the offix-scheduler implementation and not in the client itself.

I think there are two potential approaches.

1. Queue every operation and then attempt to execute operations immediately if online. This will possibly bring some challenges around the queue implementation. We'll probably need to add some locking/semaphore type behavior on the queue to ensure operations are sequenced in a predictable way.

2. Don't queue everything, but allow the scheduler to take a user defined error handler for requeueing on specific errors.

I think with either of these solutions, we'll need a way mark certain operations as stale/expired to prevent the queue from getting stuck on faulty operations. Number of retries sounds good to me

**Comment by wtrocki** [comment]
For approach nr1 it might have certain implications. For offline operations, we can guarantee no data loss as it will always save item to db before even making the request. Adding all to the queue will mean that we will need to disable ability to have parallel network requests (as this will allow updates happening before create) etc. We would also not be able to perform request without saving item to the database.

This problem can be addressed in two phases which directly relate to option 1 and 2 proposed by @darahayes.

We might start with 2 with some eventual consistency. Challenge here is that retry link might not be the best implementation here. Ideally we should not block promise till all retries are done - we should return on first failure and try to retry using offline queue - but that also has implications.

As for approach nr 1 it sounds like something that will completely invalidate current architecture/scheduler. This approach will probably require us to employ async replication model rather than instant queries and mutations that we have at the moment. If we start blocking promisses and retrying over the time entire api will become hard to use as you would not be able to do progress bar or error message with blocked UI until number of tries is exausted.


I think eventually we going to have a model where we do not have an apollo cache and offline store, but the fully-fledged database that will contain local items and asynchronous model. Sucessful promise in this case would mean that we succedded to add item to the local db and then instantly we going to give people access to another promise that has network error. This is complete shift from paradigm used in current graphql clients and it will mean that there will be not much sense to use Apollo or URQL client at this point.  

With that model, we will simply have flags like 
`_replicated=false, _deleted=false, _no_of_tries=1, sequential=false` etc. that will give us information and API that will control events that are happening and automatically garbage collect the data.

**Comment by kingsleyzissou** [comment]
Hi @skylarkgit, we have made the decision to move away from the Apollo client and we have decided to deprecate our Offix packages and release our Datastore. I will be closing this issue, but if you have any questions, please feel free to let us know.

You can see the docs for the updated datastore here:
https://offix.dev/docs/getting-started
